### DISCUSSION TITLE
How do I store an array of strings in independent memory? Add CD74HC4067 module.

### DISCUSSION THREAD

**Discussion by dejavy** [arendst/Tasmota#10786] [Q&A] [answered]
I am adding support for the new CD74HC4067 module (many asked about it, many need it)
He works for me (80%).
![1612134385043](https://user-images.githubusercontent.com/695472/106401037-8eac6980-642a-11eb-8c86-83daaf7bfb10.jpg)

I would like to add the ability to save friendly names for individual channels.
![1612134794549](https://user-images.githubusercontent.com/695472/106401057-adaafb80-642a-11eb-9931-5806dbe9c4f2.jpg)

I need help finding free space to save the array. The written code is not well documented, I don’t even know where to find a place in memory.

![1612134323428](https://user-images.githubusercontent.com/695472/106401074-ce735100-642a-11eb-80b9-6d70af4e0bad.jpg)

**Comment by sfromis** [comment, marked as answer]
150 bytes would be a lot to ask for in the 4K settings block. One option used by some drivers is to have such in the RAM for the driver, meaning that boot rules are needed to set the values again each boot. Can work quite well.

More generally, the typical way for configuring Tasmota drivers is to have commands for doing settings, with a web UI being a "secondary" interface (often not even created) not suitable for automatic updates of settings.

**Reply by dejavy** [reply]
Yes, I understand that.
But why in the save block the config is unused
#ifdef USE_
#endif USE_
so as not to occupy the memory with unnecessary data (the module is turned off, and the memory for saving the settings is occupied)

**Reply by dejavy** [reply]
I need these names for the MQTT mailing list.
There will be a large number of these devices (modules) in one network.
For clear tracking, it would be reasonable to have and receive clear names of the same sensors (when there can be 50 or 300 of them).

**Reply by dejavy** [reply]
Can I free up memory space by locking (commenting out) exactly unused parameters?

**Reply by ascillato** [reply]
> Can I free up memory space by locking (commenting out) exactly unused parameters?

If you do that, the memory map will change and your firmware will be backwards incompatible making the settings area from an update to scramble and left the device in an unoperable state. In this cases, Tasmota fallbacks to restore the config to default, deleting the user's settings. So, not a good idea to scramble the memory settings area.

Is not very optimum idea to be store just text in settings. It will take out the opportunity for more features to be added to Tasmota.

Seems that the best option is to send that data by mqtt with just an index and let the home automation software to assign user selected names.

After that, other option is to use rules and MEMs to send the data in the format that the user wants.

**Reply by sfromis** [reply]
Assuming that your module driver will be an optional one only enabled for users actually wanting to use that device (enabled or disabled by a USE_xxx #define option), allocating static RAM for the purpose may not be so bad, as it would be feasible to just avoid having a lot of other features needing RAM included at the same time.

Another option, depending on the hardware, is to use a very new feature, for now mainly experimental and in test, of using a file system to store data outside of the regular 1 megabyte memory model, assuming that a larger flash chip is present. I'm not the one to ask for details of how to do that, but AFAIK it is being used for recent BLE sensor development work.

For compatibility reasons, I agree fully with ascillato about not restructuring the settings block to cram your data into that structure.

**Comment by dejavy** [comment]
The diagram is shown here.
![1612167022994](https://user-images.githubusercontent.com/695472/106431553-1bc8e000-6476-11eb-8228-ef91e11bcf78.jpg)

She is not entirely clear to me.
Can I find 150 bytes free to store the text I need?
Or is it absolutely impossible?

**Comment by arendst** [comment]
I hate this urge to have friendlynames in Tasmota as there is simply no space to support it. Names should be provided by the Home Automation tool based on simple and short MQTT messages. Apparently it also needs a GUI page which makes things even worse. It is nice for your use-case but doesn't adhere to any basic Tasmota rule of making code extensions as small as possible while adding functionality.

But if you insist....

You cannot safely store anything outside the fixed Settings area as it is overwritten at OTA time. An alternative is using the recently released filesystem.

To use part of the Settings area I will investigate the possibility to reduce space for Rule3 and make room  for a single blob.

If the filesystem is enabled on preferrably a system with at least 4M flash you could use it to store a blob with your data. As every file takes at least 4k due to LittleFS minimal sector size it is no option to store every string as a single file hence it needs to be a blob.

In both situation you will have to write code for default names if the settings are not yet initialized. If using the filesystem there will also need to be code to detect changes in the friendlynames and asynchronisly write these changes back without wearing out flash. So only write if changes are detected.

**Comment by dejavy** [comment]
Let me put the question differently.
I can add a condition in settings.h
#ifdef USE_
#endif USE_,
For example for these modules?

  uint16_t knx_physsical_addr;
  uint16_t knx_GA_addr [MAX_KNX_GA];
  uint16_t knx_CB_addr [MAX_KNX_CB];
  uint8_t knx_GA_param [MAX_KNX_GA];
  uint8_t knx_CB_param [MAX_KNX_CB];
  
  uint8_t shutter_invert [MAX_SHUTTERS]; // E78
  uint8_t shutter_set50percent [MAX_SHUTTERS]; // E7C
  uint8_t shutter_position [MAX_SHUTTERS]; // E80
  uint8_t shutter_startrelay [MAX_SHUTTERS]; // E84

If the firmware does not use them, then it does not take up memory.

**Reply by ascillato** [reply]
@dejavy 

If you move the place of any setting, you are scrambling the settings area, so all positions changes. In that case, what was a byte for a setting is going to be another thing totally different. So your idea is backwards incompatible.

@arendst 

Please, don't reduce the space for rule3. Several users use it at close to full space.

**Reply by sfromis** [reply]
What you're suggesting there can work for your own purposes if you do not worry about upgrade compatibility, but not at all suitable as part of the project to be distributed to users. Better look at using the file system now introduced, or have the names populated by a boot rule.

**Comment by arendst** [comment]
Alas I see that I'm unable to auto attach any commits to these discussions.

Any way, I've made a demo how to use the filesystem to save driver specific settings. See latest commit https://github.com/arendst/Tasmota/commit/7561647dce62358aef8373519a84512f235c4fc6

All you need is the information in file ``xdrv_98_file_settings_demo.ino``. The other changes are just needed to get it working.

**Comment by dejavy** [comment]
In general, a very strange logic.
Not! I understand you - backward compatibility and everything in that sense.
But there is one BUT!
We are not talking about a "typical" configuration, but to assemble a project for a specific task (the one who does it understands what he is doing)
And therefore, "We - many of us" understand that if we configured something "round" in the firmware, then it will not work in something "square".
And if we need something square, we have to edit the configuration.

**Comment by dejavy** [comment]
It is very difficult for me to imagine that I first configured the "device" to control the relay.
And after a while I will configure this "device" as a temperature sensor and a watt-meter. and that in this case the "saved" relay settings will be useful to me. , and backward compatibility.

**Comment by dejavy** [comment]
If I do not fully understand the concept of Projects - obyastine me the value of data stored in the memory of those modules and drivers are not used in otdelnom version?
Toli is a light sensor
Toli are relay modules.
Each can be individually updated without data corruption
But (of course) you cannot swap them without changing the configuration.

**Comment by dejavy** [comment]
I may be wrong (please explain the idea to me) - but it seems to me if you change the configuration
#ifdef USE_
#endif USE_
It is logical that the block of stored memory will change ...

**Reply by sfromis** [reply]
The whole concept of the settings structure is that *nothing* changes, as it is not intended to be changed by varying what #define switches are used or not. That hex number in the comments indicate a *fixed* offset, only changed after very careful consideration and settings migration strategies. In a few cases, one driver is decided to be exclusive with another driver, allowing reuse of locations within settings, but still without changing any offsets.

Sure, the CD74HC4067 looks like a nice module to have supported in Tasmota, but if you aim for having your driver included in the Tasmota open source project, you'll need to work within how things are done in Tasmota. Part of the concept of Tasmota is that things can be swapped without redoing the configuration from scratch. You can "just" flash another binary including the drivers you need for some difference in hardware.

**Reply by meingraham** [reply]
#ifdef USE_ is not intended for adding code that will make a breaking change from option to option. It's intent is to configure a program size that will fit in the available flash memory.

If you want to implement a feature for yourself, the best approach is for you to fork the project and maintain your own specialized version of Tasmota.

**Comment by gemu2015** [comment]
this is an ideal use case for scripts, no driver needed at all

in this case for esp32 and no MQTT just to show how it works in principle:

<img width="631" alt="adc_muxer" src="https://user-images.githubusercontent.com/11647075/106575664-9e1bd780-653c-11eb-960e-f2940ca01690.png">

**Reply by pkkrusty** [reply]
@gemu2015 How adaptable is this script code to 8266? I find myself in need of this exact module for reading lithium cells.

**Reply by barbudor** [reply]
@pkkrusty Scripting-Language is available also as a feature for ESP8266. Actually it was created as the 1st scripting solution long before ESP32 and Berry
You need a compile with Scripting that is exclusive with Rules (it's one or the other).
If you don't want to compile yourself, there are unofficial binaries with scripting in https://github.com/tasmota/install/tree/main/firmware/unofficial

**Reply by pkkrusty** [reply]
Yes, I've used a bit of scripting in my data logger to SD card, but only scratched the surface. The code gemu wrote above looks mostly similar to what I've seen in the documentation, but I'm wondering if there's anything in that specific code exclusive to ESP32. 

I'll have to do some learning by doing probably.

**Reply by sfromis** [reply]
There are a few extra things for ESP32 due to the extra capabilities (like Webcam or more than one ADC), but nearly everything should be the same.

**Reply by gemu2015** [reply]
> @gemu2015 How adaptable is this script code to 8266? I find myself in need of this exact module for reading lithium cells.

the only cmd you have to adapt is the get adc cmd since 8266 has only one.

mux[mcnt+1]=adc(4)

**Reply by pkkrusty** [reply]
super, thanks so much
